### PR TITLE
feat: add parseISODuration

### PR DESCRIPTION
**Pull request by matinzd**
Related: https://github.com/date-fns/date-fns/issues/3150

### CODE REVIEW

**Comment by matinzd** [comment]
@leshakoss @kossnocorp

**Comment by matinzd** [comment]
@leshakoss @kossnocorp

**Comment by alfaproject** [comment]
Is there any hope to get this in?

**Comment by matinzd** [comment]
> Is there any hope to get this in?

I messaged @kossnocorp, it seems that they are waiting for v3.0.0 release so this can be included there if there won't be any bugs.

**Comment by alfaproject** [comment]
Perhaps we can re-target for v2 instead of main branch?

**Comment by kossnocorp** [comment]
If you need to use the function, please copy-paste the code to your project locally. It's not worth it for us to rebuild the function for v2.

**Comment by alfaproject** [comment]
Alright, fair enough

**Comment by alex-collibra** [comment]
First of all, thanks a lot for this workaround.
It seems like parsing of negative durations doesn't work.
```
parseISODuration('P-1D');

// results in Invalid format
```
Is it on purpose? Can this be added to this feature as well? It is supported in libs like luxon or dayjs, so technically it may be possible. Here is an example of luxon and dayjs if you want live example:
https://codesandbox.io/s/dazzling-liskov-4dqzg5?file=/src/App.tsx

**Comment by matinzd** [comment]
> First of all, thanks a lot for this workaround. It seems like parsing of negative durations doesn't work.
> 
> ```
> parseISODuration('P-1D');
> 
> // results in Invalid format
> ```
> 
> Is it on purpose? Can this be added to this feature as well? It is supported in libs like luxon or dayjs, so technically it may be possible. Here is an example of luxon and dayjs if you want live example: https://codesandbox.io/s/dazzling-liskov-4dqzg5?file=/src/App.tsx

It seems negative offsets are available in the [specification](https://en.wikipedia.org/wiki/ISO_8601). I will try to update the PR accordingly.

**Comment by matinzd** [comment]
> First of all, thanks a lot for this workaround. It seems like parsing of negative durations doesn't work.
> 
> ```
> parseISODuration('P-1D');
> 
> // results in Invalid format
> ```
> 
> Is it on purpose? Can this be added to this feature as well? It is supported in libs like luxon or dayjs, so technically it may be possible. Here is an example of luxon and dayjs if you want live example: https://codesandbox.io/s/dazzling-liskov-4dqzg5?file=/src/App.tsx

According to the [specification](https://en.wikipedia.org/wiki/ISO_8601#:~:text=Durations%20define%20the,number%20of%20seconds.) all these values are positive and above zero. In my understanding I would assume we should not have a negative day or negative year?

**Comment by alex-collibra** [comment]
> > First of all, thanks a lot for this workaround. It seems like parsing of negative durations doesn't work.
> > ```
> > parseISODuration('P-1D');
> > 
> > // results in Invalid format
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is it on purpose? Can this be added to this feature as well? It is supported in libs like luxon or dayjs, so technically it may be possible. Here is an example of luxon and dayjs if you want live example: https://codesandbox.io/s/dazzling-liskov-4dqzg5?file=/src/App.tsx
> 
> According to the [specification](https://en.wikipedia.org/wiki/ISO_8601#:~:text=Durations%20define%20the,number%20of%20seconds.) all these values are positive and above zero. In my understanding I would assume we should not have a negative day or negative year?

Agree, but it was the same discussion in other libraries (sorry to link to other libs from here, but this is the one of many discussions: https://github.com/moment/moment/issues/2408#issue-85650097)

**Comment by cj-christoph-gysin** [comment]
Any chance to get the conflicts resolved so we can still get this into v3?

File: src/parseISODuration/index.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * //=>
+ * {
+ *   years: 3,
+ *   months: 6,
+ *   weeks: 0,
+ *   days: 4,
+ *   hours: 12,
+ *   minutes: 30,
+ *   seconds: 5,
+ * }
```
Comment (kossnocorp) [inline src/parseISODuration/index.ts:36]: Please format the whole object as comments otherwise the code highlight on the website will break

File: src/parseISODuration/index.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  return Object.entries({
+    years,
+    months,
+    weeks,
+    days,
+    hours,
+    minutes,
+    seconds,
+  }).reduce((prev, [key, value]) => {
+    return { ...prev, [key]: Number(value) }
```
Comment (kossnocorp) [inline src/parseISODuration/index.ts:68]: It will be faster to mutate the object:

```suggestion
    prev[key] = +value;
    return prev;
```

**Review by kossnocorp — CHANGES_REQUESTED** [review]
Thank you a lot for your work and waiting! I finally have time to work on new stuff and will be happy to ship it with the next version

**Comment by kossnocorp** [comment]
Please rebase and reformat code with the default Prettier config, thanks!

File: src/parseISODuration/index.ts (hunk 2 above)
Comment (matinzd) [reply src/parseISODuration/index.ts]: Fixed [320c1f9](https://github.com/date-fns/date-fns/pull/3151/commits/320c1f9862a1003b751eb04fc88d993c83b53d09).

File: src/parseISODuration/index.ts (hunk 1 above)
Comment (matinzd) [reply src/parseISODuration/index.ts:36]: Fixed [320c1f9](https://github.com/date-fns/date-fns/pull/3151/commits/320c1f9862a1003b751eb04fc88d993c83b53d09).

File: src/parseISODuration/index.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ * Parse the given string in ISO 8601 format and return Duration object.
+ *
+ * Function accepts duration format of ISO 8601.
+ * ISO 8601: http://en.wikipedia.org/wiki/ISO_8601
+ * Specifically: https://en.wikipedia.org/wiki/ISO_8601#Durations
+ *
+ * If the argument is not a valid ISO Duration it will throw Invalid format error.
+ *
+ * @param argument - ISO Duration string
+ * @returns duration object
```
[the remark this answers is not in the document]
Comment (matinzd) [reply src/parseISODuration/index.ts]: Fixed [320c1f9](https://github.com/date-fns/date-fns/pull/3151/commits/320c1f9862a1003b751eb04fc88d993c83b53d09).

File: src/parseISODuration/index.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * @throws {RangeError} Duration in ISO format cannot be `Invalid format`
```
[the remark this answers is not in the document]
Comment (matinzd) [reply src/parseISODuration/index.ts]: Fixed [320c1f9](https://github.com/date-fns/date-fns/pull/3151/commits/320c1f9862a1003b751eb04fc88d993c83b53d09).

**Comment by matinzd** [comment]
> Please rebase and reformat code with the default Prettier config, thanks!

Rebased now and all suggestions are applied. However I don't know how to test generated docs to make sure if the example works properly. Let me know if that still has an issue.

**Comment by matinzd** [comment]
I guess we may need to add more test cases as I have heard a lot about negative values. WDYT? @kossnocorp

**Comment by kossnocorp** [comment]
@matinzd don't worry about the docs. Adding more tests never hurt, if you feel there's something to cover please go ahead.

**Comment by andrioid** [comment]
What is missing so this can be merged? Anything I can do to help?

**Comment by matinzd** [comment]
@andrioid 
There are some test cases that the current implementation is not covering. I will try to update the PR accordingly.

**Comment by matinzd** [comment]
@yjukaku Yes, you are right! The test case I found was not related to supporting negative values. Also negative values are not in the standard as far as I checked. 
I really want to be mindful when merging PR because I don't want people to use broken function in their codebase especially that date-fns is widely used in the ecosystem.

File: src/parseISODuration/index.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ * //=>
+ * const result = {
+ *   years: 3,
+ *   months: 6,
+ *   weeks: 0,
+ *   days: 4,
+ *   hours: 12,
+ *   minutes: 30,
+ *   seconds: 5,
+ * }
```
Comment (ckcherry23) [inline src/parseISODuration/index.ts:36]: @matinzd I think what we need here is for the entire result object to be formatted as a comment. You can refer to `src/intervalToDuration/index.ts` for an example.

```suggestion
 * //=> { years: 3, months: 6, weeks: 0, days: 4, hours: 12, minutes: 30, seconds: 5 }
```

**Review by ckcherry23 — COMMENTED** [review]
Hi @matinzd, can you let us know what test cases you think can be added before we merge this in? I can help to contribute as well.

**Review by ckcherry23 — COMMENTED** [review]
Hi @matinzd, maybe we could also add more invalid test cases for this, for example:
* `P` - only P
* `PT` - only P and T
* `P1H` - missing T
* `PT1D` - unneeded T
* `P1W1YT1M` - incorrect date order before T
* `P1WT1M1H` - incorrect time order after T

We could also have:
* `P0D` - gives zero instead of undefined
* `PT0S`- gives zero instead of undefined
* `PT0.0021S` - 0.0021S (instead of 2.1 milliseconds, which could be useful for `ExtendedDuration` if supported later)
